### PR TITLE
BREAKING CHANGE: Upgrade Spring to 7.x, Spring Boot to 4.x

### PR DESCRIPTION
**Pull request by JordanJLopez**
### :pencil: Description

NOTE: This PR is built off of #2147, and is a superset of that one. That one should be merged first.

This pull requests upgrades to the latest Spring Boot and Spring versions.

For reference, see https://github.com/spring-projects/spring-boot/wiki/Spring-Boot-4.0-Migration-Guide

The change with the biggest impact is the splitting up of certain modules within the Spring Boot ecosystem, requiring that we pull more specific dependencies.

#### Jackson Upgrade Delayed

https://github.com/spring-projects/spring-boot/wiki/Spring-Boot-4.0-Migration-Guide#upgrading-jackson

Spring Boot 4 would prefer that we move to Jackson 3, but we're blocked on that while graphql-java remains on Jackson 2. Therefore we need to ensure that we use the Spring Boot Jackson2 module.

**Additionally,** some tests broke when Jackson 3 was included in the classpath, due to an internal Spring flag that checks the classpath: https://github.com/spring-projects/spring-framework/blob/77bc13dc8c4387ad33af274eed39ed2734edce56/spring-websocket/src/main/java/org/springframework/web/socket/sockjs/transport/TransportHandlingSockJsService.java#L74

(The above link indicates the kind of logic, I'm not entirely sure it's the exact spot that was breaking the test).

### CODE REVIEW

File: examples/federation/products-subgraph/src/test/kotlin/com/expediagroup/graphql/examples/federation/products/ProductsApplicationTest.kt [hunk 1]
```diff
@@ -17,16 +17,23 @@
...
 import org.springframework.beans.factory.annotation.Autowired
-import org.springframework.boot.test.autoconfigure.web.reactive.AutoConfigureWebTestClient
 import org.springframework.boot.test.context.SpringBootTest
+import org.springframework.context.ApplicationContext
 import org.springframework.http.MediaType
 import org.springframework.test.web.reactive.server.WebTestClient
 
 @SpringBootTest
-@AutoConfigureWebTestClient
-class ProductsApplicationTest(@Autowired val testClient: WebTestClient) {
```
Comment (JordanJLopez) [inline examples/federation/products-subgraph/src/test/kotlin/com/expediagroup/graphql/examples/federation/products/ProductsApplicationTest.kt:29]: You'll see this elsewhere, but this is a result of the splitting up of Spring modules.
https://github.com/spring-projects/spring-boot/wiki/Spring-Boot-4.0-Migration-Guide#using-webclient-or-testresttemplate-and-springboottest

Followed this guide for the new way of setting up the test client https://docs.spring.io/spring-framework/reference/testing/webtestclient.html

File: gradle/libs.versions.toml [hunk 2]
```diff
@@ -82,9 +81,14 @@ maven-plugin-api = { group = "org.apache.maven", name = "maven-plugin-api", vers
 maven-project = { group = "org.apache.maven", name = "maven-project", version.ref = "maven-project" }
 poet = { group = "com.squareup", name = "kotlinpoet", version.ref = "poet" }
 slf4j = { group = "org.slf4j", name = "slf4j-api", version.ref = "slf4j" }
+spring-boot-autoconfigure = { group = "org.springframework.boot", name = "spring-boot-autoconfigure", version.ref = "spring-boot" }
```
Comment (JordanJLopez) [inline gradle/libs.versions.toml:84]: New dependencies from the module splitting https://github.com/spring-projects/spring-boot/wiki/Spring-Boot-4.0-Migration-Guide#module-dependencies

File: gradle/libs.versions.toml [hunk 3]
```diff
@@ -82,9 +81,14 @@ maven-plugin-api = { group = "org.apache.maven", name = "maven-plugin-api", vers
...
 spring-boot-config = { group = "org.springframework.boot", name = "spring-boot-configuration-processor", version.ref = "spring-boot" }
 spring-boot-netty = { group = "org.springframework.boot", name = "spring-boot-starter-reactor-netty", version.ref = "spring-boot" }
 spring-boot-webflux = { group = "org.springframework.boot", name = "spring-boot-starter-webflux", version.ref = "spring-boot" }
+spring-boot-webflux-module = { group = "org.springframework.boot", name = "spring-boot-webflux", version.ref = "spring-boot" }
+# TODO: Migrate to Jackson 3.x when graphql-java upgrades
+# spring-boot-jackson = { group = "org.springframework.boot", name = "spring-boot-jackson", version.ref = "spring-boot" }
+spring-boot-jackson2 = { group = "org.springframework.boot", name = "spring-boot-jackson2", version.ref = "spring-boot" }
```
Comment (JordanJLopez) [inline gradle/libs.versions.toml:91]: Spring would prefer we move to Jackson 3, but we need graphql-java to get there first.

https://github.com/spring-projects/spring-boot/wiki/Spring-Boot-4.0-Migration-Guide#jackson-2-compatibility

File: integration/graalvm/ktor-graalvm-server/build.gradle.kts [hunk 4]
```diff
@@ -33,6 +33,10 @@ graalvmNative {
         named("main") {
             verbose.set(true)
             buildArgs.add("--initialize-at-build-time=io.ktor,kotlin,kotlinx.io,ch.qos.logback,org.slf4j")
+            // Keep Ktor CIO pipeline initialization at runtime to avoid embedding coroutine context objects.
+            buildArgs.add("--initialize-at-run-time=io.ktor.server.cio.PipelineKt")
+            // GraalVM 25 is stricter about coroutine objects ending up in the image heap.
+            buildArgs.add("--initialize-at-run-time=kotlinx.coroutines")
```
Comment (JordanJLopez) [inline integration/graalvm/ktor-graalvm-server/build.gradle.kts:39]: The newer GraalVM version (we're going to 25 in the tests) is stricter about these runtime classes.

File: servers/graphql-kotlin-spring-server/src/test/kotlin/com/expediagroup/graphql/server/spring/routes/RouteConfigurationIT.kt [hunk 5]
```diff
@@ -275,6 +285,7 @@ class RouteConfigurationIT(@Autowired private val testClient: WebTestClient) {
 
     @Test
     fun `verify POST graphQL request with explicit charset`() {
+        val utf8Json = MediaType.parseMediaType("application/json;charset=UTF-8")
```
Comment (JordanJLopez) [inline servers/graphql-kotlin-spring-server/src/test/kotlin/com/expediagroup/graphql/server/spring/routes/RouteConfigurationIT.kt:288]: `MediaType.APPLICATION_JSON_UTF8` constant was deprecated and removed. To continue testing this edge case, we need to manually recreate it.

File: servers/graphql-kotlin-spring-server/build.gradle.kts [hunk 6]
```diff
@@ -9,19 +23,35 @@ plugins {
 dependencies {
     api(projects.graphqlKotlinServer)
     api(projects.graphqlKotlinFederation)
-    api(libs.spring.boot.webflux)
+    api(libs.spring.boot.webflux.module)
+    api(libs.spring.boot.autoconfigure)
+    api(libs.spring.boot.netty)
+    api(libs.spring.boot.jackson2)
```
Comment (JordanJLopez) [inline servers/graphql-kotlin-spring-server/build.gradle.kts:29]: Spring Boot modularization

File: servers/graphql-kotlin-spring-server/build.gradle.kts [hunk 7]
```diff
@@ -37,7 +67,7 @@ tasks {
             }
         }
     }
-    val dokkaJavadoc = getByName("dokkaJavadoc") {
+    named("dokkaGeneratePublicationHtml") {
```
Comment (JordanJLopez) [inline servers/graphql-kotlin-spring-server/build.gradle.kts:70]: As part of this upgrade, we moved to Dokka v2's html doc publication.

The Dokka v2 javadoc support is in ["alpha"](https://kotlinlang.org/docs/dokka-migration.html#select-documentation-output-format), and they don't guarantee that it will be usable by other tools. I felt it was safer to with the new default.

More info on the dokka migration here https://kotlinlang.org/docs/dokka-migration.html

**Review by tapaderster — APPROVED** [review]
